### PR TITLE
Develop andrew

### PR DESCRIPTION
**Pull request by andrewljohnson**

### CODE REVIEW

File: js/cardview.js [hunk 1]
```diff
@@ -6,6 +6,7 @@ let Card = React.createClass({
 
   // layout and style the card
   render() {
+    console.log(this.props)
```
Comment (lacker) [inline js/cardview.js:9]: Do you need to keep this in? I vote remove

File: js/client.js [hunk 2]
```diff
@@ -60,20 +61,37 @@ class Client {
   }
 
   // Send a looking-for-game message.
-  register() {
+  register(hasComputerOpponent) {
```
Comment (lacker) [inline js/client.js:64]: i support this way of doing it

File: js/client.js [hunk 3]
```diff
@@ -60,20 +61,37 @@ class Client {
...
     this.registered = true; // UI checks this to refresh on game seek
     console.log("registering as " + this.name)
     if (this.game) {
-      this.send({op: "register", name: this.name, seeking: !this.game.started()})
+      this.send({op: "register", name: this.name, seeking: !this.game.started(), hasComputerOpponent:hasComputerOpponent})
```
Comment (lacker) [inline js/client.js:68]: you might just be able to do

this.send({op: "register", name: this.name, seeking: !this.game.started(), hasComputerOpponent})

depending on whether that subset of es6 is supported. i'd try and see

File: js/client.js [hunk 4]
```diff
@@ -105,6 +123,13 @@ class Client {
     this.send(move)
   }
 
+  makeComputerMove(move) {
```
Comment (lacker) [inline js/client.js:127]: can you put an enlightening comment here?

File: js/gamestate.js [hunk 5]
```diff
@@ -16,15 +16,16 @@ class PlayerState {
     this.maxMana = data.maxMana || 0
   }
   
-  // Creates a string that, when printed, is a nice way to view the
-  // contents of the PlayerState for debugging.
+  // A string that can be displayed to debug the game state.
+  // This string should be consistent for all clients viewing the same  
+  // game. In particular, it prints players by alphabetical order.
```
Comment (lacker) [inline js/gamestate.js]: eh, this comment doesn't appear to be correct here? what did you do?

File: js/gamestate.js [hunk 6]
```diff
@@ -103,10 +112,9 @@ class GameState {
     return answer
   }
 
-  // A string that can be displayed to debug the game state.
-  // This string should be consistent for all clients viewing the same
-  // game. In particular, it prints players by alphabetical order.
-  displayString() {
+  // Creates a string that, when printed, is a nice way to view the
```
Comment (lacker) [inline js/gamestate.js]: did you mangle this comment?

File: js/gamestate.js [hunk 7]
```diff
@@ -226,14 +234,33 @@ class GameState {
...
+    
+    this.checkForWinner()
+  }
 
-    if (this.current().life <= 0) {
-      this.winner = this.opponent().name
-    } else if (this.opponent().life <= 0) {
-      this.winner = this.current().name
+  // set the winner and trigger some animation or show if game ends
+  checkForWinner () {
```
Comment (lacker) [inline js/gamestate.js]: no space before ()

File: js/gamestate.js [hunk 8]
```diff
@@ -226,14 +234,33 @@ class GameState {
...
       console.log(this.winner + " wins!")
+      alert(this.winner + " wins!")
+      this.declaredWinner = true
+      
+      // stop all cards from attacking
+      for (var i = 0; i < this.players.length; i++) {
+        var player = this.players[i]
+        for (var j = 0; j < player.board.length; j++) {
+          var card = player.board[j]
+          if (card.attackLoop) {
```
Comment (lacker) [inline js/gamestate.js:258]: something feels awry here

File: js/gamestate.js [hunk 9]
```diff
@@ -382,12 +410,28 @@ class GameState {
...
+    // for permanents that attack on a loop
+    if (card.attackRate) {
+      // default to attack opponent's face
+      for (let p of this.players) {
+        if (p == player) {          
+          card.attacker = p
+          break;
+        }
+      }
+      card.attackLoop = setInterval(() => {
```
Comment (lacker) [inline js/gamestate.js:422]: the whole attackLoop logic, i'm afraid there's paralellism bugs in it

File: js/client.js (hunk 3 above)
Comment (andrewljohnson) [inline js/client.js:68]: :checkmark:

File: js/gamestate.js (hunk 5 above)
Comment (andrewljohnson) [inline js/gamestate.js]: worry not

File: js/gamestate.js (hunk 6 above)
Comment (andrewljohnson) [inline js/gamestate.js]: ftfy

File: js/gamestate.js (hunk 8 above)
Comment (andrewljohnson) [inline js/gamestate.js:258]: pls explain?

**Comment by lacker** [comment]
LGTM

**Pull request merged**